File: python/S2_SORTING/L2_ADVANCED/quicksort.py
class Sorting:
    def partition(self,arr,low,high):
        pivot=arr[high]
        i=low-1
        for j in range(low,high):
            if arr[j]<=pivot:
                i+=1
                arr[i],arr[j]=arr[j],arr[i]
        arr[i+1],arr[high]=arr[high],arr[i+1]
        return i+1
    def quicksort(self,arr,low,high):
        if low<high:
            p=self.partition(arr,low,high)
            self.quicksort(arr,low,p-1)
            self.quicksort(arr,p+1,high)

if __name__=="__main__":
    n=int(input())
    arr=list(map(int,input().split()))
    s=Sorting()
    s.quicksort(arr,0,n-1)
    for i in range(0,n):
        print(arr[i],end=" ")
        

File: C/S2_SORTING/L2_ADVANCED/quicksort.c
/*
Quick Sort

Algorithm:
Select a pivot element.
Partition the array such that:
- Elements smaller than pivot go to the left
- Elements greater than pivot go to the right
Recursively apply the same process to left and right subarrays.

Time Complexity:
Worst Case: O(n^2)     // when pivot is always smallest/largest (bad choice)
Average Case: O(n log n)
Best Case: O(n log n)

Space Complexity:
O(log n)   // due to recursion stack (in-place apart from this)

Properties:
- In-place sorting (with small recursion stack)
- Not stable
*/
#include<stdio.h>
int partition(int arr[],int l,int r){
    int pivot=arr[r];
    int i=l-1;
    for(int j=l;j<r;j++){
        if(arr[j]<=pivot){
            i++;
            int t=arr[i];
            arr[i]=arr[j];
            arr[j]=t;
        }
    }
    int t=arr[i+1];
    arr[i+1]=arr[r];
    arr[r]=arr[i+1];
    return i+1;
}
void quicksort(int arr[],int l,int r){
    if(l<r){
        int p=partition(arr,l,r);
        quicksort(arr,l,p-1);
        quicksort(arr,p+1,r);
    }
}

int main(){
    int n;
    printf("Enter size of array: ");
    scanf("%d",&n);
    int arr[n];
    printf("Enter element of array: ");
    for(int i=0;i<n;i++){
        scanf("%d",&arr[i]);
    }
    quicksort(arr,0,n-1);
    for(int i=0;i<n;i++){
        printf("%d ",arr[i]);
    }
    return 0;
}